d2/src/main/java/com/linkedin/d2/balancer/servers/ZooKeeperAnnouncer.java

```diff
@@ -93,6 +98,23 @@ public class ZooKeeperAnnouncer implements D2ServiceDiscoveryEventHelper
   private final AtomicLong _markDownStartAtRef = new AtomicLong(Long.MAX_VALUE);
 
   private volatile Map<Integer, PartitionData> _partitionDataMap;
+  /**
+   * If not null, it defines two rules for d2 weight validation:
+   * 1. The maximum d2 weight allowed.
+   * 2. The maximum number of decimal places allowed. Use 0s on decimal places to indicate it.
+   * For example, 100.00 means the max weight allowed is 100 and the max number of decimal places is 2.
+   * CAUTION: BigDecimal yields accurate scale when constructed with a string of the number, instead of a double/float.
+   * E.g: new BigDecimal("100.00") instead of new BigDecimal(100.00).
+   */
+  private final BigDecimal _maxWeight;
+  /**
+   * The action to take when d2 weight breaches validation rules.
+   */
+  private ActionOnWeightBreach _actionOnWeightBreach = ActionOnWeightBreach.IGNORE;
```

Comment:
If you write the code like this in the constructor, this can be upgraded to `final` instead:
```java
    _actionOnWeightBreach = (actionOnWeightBreach != null) ? actionOnWeightBreach : IGNORE;
```
Just makes it a little cleaner